Python/legacy_tracing.c

```diff
@@ -0,0 +1,541 @@
+        Py_RETURN_NONE;
+    }
+    assert(PyVectorcall_NARGS(nargsf) == 2);
+    int line = _PyLong_AsInt(args[1]);
+    assert(line >= 0);
+    PyFrameObject* frame = PyEval_GetFrame();
+    if (frame == NULL) {
+        PyErr_SetString(PyExc_SystemError,
+                        "Missing frame when calling trace function.");
+        return NULL;
+    }
+    assert(args[0] == (PyObject *)frame->f_frame->f_code);
+    if (frame ->f_last_traced_line == line) {
+        /* Already traced this line */
+        Py_RETURN_NONE;
+    }
+    return trace_line(tstate, self, frame, line);
+}
+
+
+static PyObject *
+sys_trace_jump_func(
+    _PyLegacyEventHandler *self, PyObject *const *args,
+    size_t nargsf, PyObject *kwnames
+) {
+    assert(kwnames == NULL);
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (tstate->c_tracefunc == NULL) {
+        Py_RETURN_NONE;
+    }
+    assert(PyVectorcall_NARGS(nargsf) == 3);
+    int from = _PyLong_AsInt(args[1]);
+    assert(from >= 0);
+    int to = _PyLong_AsInt(args[2]);
+    assert(to >= 0);
+    PyFrameObject* frame = PyEval_GetFrame();
+    if (frame == NULL) {
+        PyErr_SetString(PyExc_SystemError,
+                        "Missing frame when calling trace function.");
+        return NULL;
+    }
+    if (!frame->f_trace_lines) {
+        Py_RETURN_NONE;
+    }
+    PyCodeObject *code = (PyCodeObject *)args[0];
+    assert(PyCode_Check(code));
+    assert(code == frame->f_frame->f_code);
+    /* We can call _Py_Instrumentation_GetLine because we always set
+    * line events for tracing */
+    int to_line = _Py_Instrumentation_GetLine(code, to);
+    /* Backward jump: Always generate event
+     * Forward jump: Only generate event if jumping to different line. */
+    if (to > from && frame->f_last_traced_line == to_line) {
+        /* Already traced this line */
+        Py_RETURN_NONE;
+    }
+    return trace_line(tstate, self, frame, to_line);
+}
+
+/* We don't care about the exception here,
+ * we just treat it as a possible new line
+ */
+static PyObject *
+sys_trace_exception_handled(
+    _PyLegacyEventHandler *self, PyObject *const *args,
+    size_t nargsf, PyObject *kwnames
+) {
+    assert(kwnames == NULL);
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (tstate->c_tracefunc == NULL) {
+        Py_RETURN_NONE;
+    }
+    assert(PyVectorcall_NARGS(nargsf) == 3);
+    PyFrameObject* frame = PyEval_GetFrame();
+    PyCodeObject *code = (PyCodeObject *)args[0];
+    assert(PyCode_Check(code));
+    assert(code == frame->f_frame->f_code);
+    int offset = _PyLong_AsInt(args[1]);
+    /* We can call _Py_Instrumentation_GetLine because we always set
+    * line events for tracing */
+    int line = _Py_Instrumentation_GetLine(code, offset);
+    if (frame->f_last_traced_line == line) {
+        /* Already traced this line */
+        Py_RETURN_NONE;
+    }
+    return trace_line(tstate, self, frame, line);
+}
+
+
+PyTypeObject _PyLegacyEventHandler_Type = {
```

Comment:
This should be added to https://github.com/python/cpython/blob/47753ecde21b79b5c5f11d883946fda2a340e427/Objects/object.c#L1982

Reply:
Why?

Shared static objects need to be stateless (and ideally immortal and immutable once we have immortal objects).

There can be some very minimal state provided that initializing it is idempotent and thread-safe, but that should not be responsibility of the class, but the interpreter.

Reply:
See https://github.com/python/cpython/blob/main/Objects/typeobject.c#L7012, it seems that `_PyStaticType_InitBuiltin`, called from `_PyTypes_InitTypes` in object.c for every type in that array, finishes initializing types (not ideal, but that's the current convention).